apps/pkcs12.c

```diff
@@ -311,13 +308,38 @@ int pkcs12_main(int argc, char **argv)
         case OPT_ENGINE:
             e = setup_engine(opt_arg(), 0);
             break;
+        case OPT_LEGACY_ALG:
+            use_legacy = 1;
+            break;
         case OPT_PROV_CASES:
             if (!opt_provider(o))
                 goto end;
             break;
         }
     }
     argc = opt_num_rest();
+
+    if (use_legacy) {
+        /* load the legacy provider if not loaded already*/
+        if (!OSSL_PROVIDER_available(app_get0_libctx(), "legacy")) {
+            if (!app_provider_load(app_get0_libctx(), "legacy"))
+                goto end;
+        }
+
+        /* load the default provider explicitly */
+        if (!app_provider_load(app_get0_libctx(), "default"))
+            goto end;
```

Comment:
This should probably be moved inside the previous if block.

Reply:
default provider won't get loaded then if legacy provided was somehow already found through OSSL_PROVIDER_available() right?

Reply:
I would say that unless we load the legacy provider explicitly, we do not want to mess with the provider loading at all. User might have configuration where both legacy and default is loaded already. If he has a (broken) configuration where the default is not loaded alongside the legacy provider then he deserves to get the failure from the missing crypto from the default provider anyway.